### PR TITLE
ROE-1447 sanitise exceptions in global exception handler

### PR DESCRIPTION
**Pull request by markpit**
https://companieshouse.atlassian.net/browse/ROE-1447

The GlobalExceptionHandler will now encode and truncate exception messages and stack traces so that if any json parse errors occur, the original value will be encoded and the stack trace truncated before logging.

### CODE REVIEW

File: src/main/java/uk/gov/companieshouse/overseasentitiesapi/exception/GlobalExceptionHandler.java [hunk 1]
```diff
@@ -1,22 +1,44 @@
...
 
+    @Value("${GLOBAL_EXCEPTION_HANDLER_TRUNCATE_LENGTH_CHARS:15000}")
+    private int truncationLength;
+
     @ExceptionHandler(Exception.class)
-    public ResponseEntity<Object> handleException(Exception ex) {
+    public ResponseEntity<Object> handleException(Exception ex, WebRequest webRequest) {
+        var context = webRequest.getHeader(ERIC_REQUEST_ID_KEY);
+        var sanitisedExceptionMessage = truncate(Encode.forJava(ex.getMessage()));
+        var sanitisedStackTrace = truncate(Encode.forJava(ExceptionUtils.getStackTrace(ex)));
```
Comment (mwestacott) [inline src/main/java/uk/gov/companieshouse/overseasentitiesapi/exception/GlobalExceptionHandler.java:28]: Just wonder if we should have a longer truncation length for the stack trace that may have nested exception stack traces in it? Although if it was ok with your testing in Docker and enough info showed up then that's fine.

Comment (markpit) [reply src/main/java/uk/gov/companieshouse/overseasentitiesapi/exception/GlobalExceptionHandler.java:28]: the stack traces in testing were about 11k chars, i'll double check

Comment (markpit) [reply src/main/java/uk/gov/companieshouse/overseasentitiesapi/exception/GlobalExceptionHandler.java:28]: the invalid date in managing officer individual dob generates a stack trace of 11.8k. The Identity date in due diligence generates about 11k so seems to be in the region of 11-12k.

Comment (mwestacott) [reply src/main/java/uk/gov/companieshouse/overseasentitiesapi/exception/GlobalExceptionHandler.java:28]: 👍 all good then.

**Review by mwestacott — APPROVED** [review]

**Pull request merged**